File: goimp/patterns/creational.go
package patterns

import (
	"fmt"

	"goimp/components"
)

func RunAbstractFactory() {
	m1 := components.CreateGame(components.MazeFactory{})
	fmt.Printf("maze 1: %v\n", m1)

	m2 := components.CreateGame(components.EnchantedMazeFactory{})
	fmt.Printf("maze 2: %v\n", m2)

	m3 := components.CreateGame(components.BombedMazeFactory{})
	fmt.Printf("maze 3: %v\n", m3)
}

func RunBuilder() {
	m1 := components.BuildGame(&components.SimpleMazeBuilder{})
	fmt.Printf("maze 1: %v\n", m1)

	m2 := components.BuildGame(&components.StandardMazeBuilder{})
	fmt.Printf("maze 2: %v\n", m2)

	b := &components.CountingMazeBuilder{}
	components.BuildGame(b)
	rooms, doors := b.GetCounts()
	fmt.Printf("maze rooms:%v, doors: %v\n", rooms, doors)
}

func RunFactoryMethods() {
	m1 := components.MakeSimpleGame().Create()
	fmt.Printf("maze 1: %v\n", m1)

	m2 := components.MakeBombedGame().Create()
	fmt.Printf("maze 2: %v\n", m2)

	m3 := components.MakeEnchantedGame().Create()
	fmt.Printf("maze 3: %v\n", m3)
}

File: py/patterns/creational.py
from .components.builder import MazeBuilder, StandardMazeBuilder, CountingMazeBuilder
from .components.games import MazeGame, create, build, EnchantedMazeGame, BombedMazeGame
from .components.factories import MazeFactory, EnchantedMazeFactory, BombedMazeFactory


def run_abstract_factory():
    print("starting the game")
    game = MazeGame()

    m1 = create(MazeFactory())
    print(m1)

    m2 = create(EnchantedMazeFactory())
    print(m2)

    m3 = create(BombedMazeFactory())
    print(m3)


def run_builder():
    game = MazeGame()
    builder1 = MazeBuilder()

    build(builder1)
    m1 = builder1.get_maze()
    print(m1)

    builder2 = StandardMazeBuilder()

    build(builder2)
    m2 = builder2.get_maze()
    print(m2)

    builder3 = CountingMazeBuilder()

    build(builder3)
    r, d = builder3.get_counts()
    print("maze 3: room count: {}, door count: {}".format(r, d))

def run_factory_method():
    m1 = MazeGame().create()
    print(m1)

    m2 = EnchantedMazeGame().create()
    print(m2)

    m3 = BombedMazeGame().create()
    print(m3)
